packages/grid/src/GridMetricCalculator.ts

```diff
@@ -1881,6 +1883,154 @@ export class GridMetricCalculator {
+    state: GridMetricState,
+    renderer: GridRenderer,
+    renderState: GridRenderState
+  ): TokenBox[] | null {
+    return this.getCachedTokenBoxesForVisibleCell(
+      column,
+      row,
+      state,
+      renderer,
+      renderState
+    );
+  }
+
+  getCachedTokenBoxesForVisibleCell = memoizeClear(
+    (
+      column: VisibleIndex,
+      row: VisibleIndex,
+      state: GridMetricState,
+      renderer: GridRenderer,
+      renderState: GridRenderState
+    ): TokenBox[] | null => {
+      const { context, model, theme } = state;
+      const metrics = this.getMetrics(state);
+
+      if (row == null || column == null || context == null || metrics == null) {
+        return null;
+      }
+
+      const modelRow = this.getModelRow(row, state);
+      const modelColumn = this.getModelColumn(column, state);
+
+      const text = model.textForCell(modelColumn, modelRow);
+      const {
+        width: textWidth,
+        x: textX,
+        y: textY,
+      } = renderer.getTextRenderMetrics(context, renderState, column, row);
+      const {
+        fontWidths,
+        gridX,
+        gridY,
+        width: gridWidth,
+        height: gridHeight,
+        verticalBarWidth,
+        horizontalBarHeight,
+      } = metrics;
+
+      // Set the font and change it back after
+      context.save();
+      context.font = theme.font;
+
+      const fontWidth =
+        fontWidths.get(context.font) ?? GridRenderer.DEFAULT_FONT_WIDTH;
+      const truncationChar = model.truncationCharForCell(modelColumn, modelRow);
+      const truncatedText = renderer.getCachedTruncatedString(
+        context,
+        text,
+        textWidth,
+        fontWidth,
+        truncationChar
+      );
+
+      const links = model.tokensForCell(
+        modelColumn,
+        modelRow,
+        truncatedText.length
+      );
+
+      // Check if the truncated text contains a link
+      if (links.length === 0) {
+        context.restore();
+        return null;
+      }
+
+      const {
+        actualBoundingBoxAscent,
+        actualBoundingBoxDescent,
+      } = context.measureText(truncatedText);
+      const linkHeight = actualBoundingBoxAscent + actualBoundingBoxDescent;
+
+      // Consider edge cases
+      const top = Math.max(gridY, gridY + textY - actualBoundingBoxAscent);
+      const bottom = clamp(
+        top + linkHeight,
+        top,
+        gridHeight - horizontalBarHeight
+      );
+
+      const tokenBoxes: TokenBox[] = [];
+
+      // Loop through the token boxes
+      for (let i = 0; i < links.length; i += 1) {
+        const { end, start } = links[i];
+        let { value } = links[i];
+
+        if (end > truncatedText.length) {
+          value = truncatedText.substring(start);
+        }
+
+        let { width: linkWidth } = context.measureText(value);
+        // Measure the width of the substring before the link
+        const startX =
+          context.measureText(truncatedText.substring(0, start)).width + textX;
```

Comment:
This is suboptimal - we're measuring the text for all the previous segments which we have already calculated. Consider the case for a string such as `one mike@deephaven.io two emily@deephaven.io three https://deephaven.io` (which gives us three tokens). Before the for loop you're measuring the whole string, then for the 0th token (`mike@deephaven.io`) you measure `mike@deephaven.io` and `one `; then for the 1st token (`emily@deephaven.io`) you're measuring `emily@deephaven.io` and `one mikebender@deephaven.io two `; then for the 2nd token (`https://deephaven.io`) you're measuring `https://deephaven.io`, then `one mike@deephaven.io two emily@deephaven.io three `. So in total you're measuring `one ` and `mikebender@deephaven.io` segments 3 times, `two ` and `emily@deephaven.io` segments twice (not including the time you measure the full string).
If you instead loop through using a `while` (as suggested in the `GridRenderer` when drawing), measuring the segments as you go through them and just storing the current width, you'll only need to measure each segment once. E.g.
```
let tokenIndex = 0;
let textStart = 0;
let textWidth = 0;
while (textStart < truncatedText.length && tokenIndex < tokens.length) {
    const token = tokens[tokenIndex];
    textWidth += context.measureText(truncatedText.substring(textStart, token.start)).width;
    const tokenWidth = context.measureText(token.value);
    ... // Calculate the box and push it to the array

    textStart = nextToken.end + 1;
    textWidth += tokenWidth;
}
```